gossip4/node.go

```diff
@@ -0,0 +1,494 @@
+package gossip4
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	cbornode "github.com/ipfs/go-ipld-cbor"
+
+	"github.com/AsynkronIT/protoactor-go/actor"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-msgio"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-hamt-ipld"
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/quorumcontrol/chaintree/nodestore"
+	"github.com/quorumcontrol/messages/v2/build/go/services"
+	"github.com/quorumcontrol/tupelo-go-sdk/bls"
+	"github.com/quorumcontrol/tupelo-go-sdk/gossip3/types"
+	"github.com/quorumcontrol/tupelo-go-sdk/p2p"
+)
+
+const gossip4Protocol = "tupelo/v0.0.1"
+
+const transactionTopic = "g4-transactions"
+
+func init() {
+	cbornode.RegisterCborType(services.AddBlockRequest{})
+}
+
+type snowballTicker struct {
+	ctx context.Context
+}
+
+type Node struct {
+	name string
+
+	p2pNode     p2p.Node
+	signKey     *bls.SignKey
+	notaryGroup *types.NotaryGroup
+	dagStore    nodestore.DagStore
+	hamtStore   *hamt.CborIpldStore
+	pubsub      *pubsub.PubSub
+
+	mempool  *mempool
+	inflight *lru.Cache
+
+	logger logging.EventLogger
+
+	rounds *roundHolder
+
+	snowballer *snowballer
+
+	signerIndex int
+
+	pid         *actor.PID
+	snowballPid *actor.PID
+	syncerPid   *actor.PID
+
+	rootContext *actor.RootContext
+}
+
+type NewNodeOptions struct {
+	P2PNode          p2p.Node
+	SignKey          *bls.SignKey
+	NotaryGroup      *types.NotaryGroup
```

Comment:
@tobowers Just to make sure I understand, the receive token tx's won't get validated because the `IsTokenRecipient` validator function isn't setup in here?

Reply:
Talked this over with @zonotope in Keybase. I have a better understanding of what needs to change and why now.